File: python/datastructure/sparseTable.py
class SparseTable:
    def __init__(self, arr):
        n = len(arr)
        k = n.bit_length()
        self.max = [[0] * n for _ in range(k)]
        for i in range(n):
            self.max[0][i] = arr[i]
        for j in range(1, k):
            for i in range(n - (1 << j) + 1):
                self.max[j][i] = max(self.max[j - 1][i], self.max[j - 1][i + (1 << (j - 1))])
    
    def query(self, l, r):
        k = (r - l + 1).bit_length() - 1
        return max(self.max[k][l], self.max[k][r - (1 << k) + 1])

# luogu 3865 【模板】ST表
if __name__ == "__main__":
    import sys
    input = lambda: sys.stdin.readline().rstrip('\n')
    print = lambda x: sys.stdout.write(str(x) + '\n')
    n, m = map(int, input().split())
    a = list(map(int, input().split()))
    st = SparseTable(a)
    for _ in range(m):
        l, r = map(int, input().split())
        print(st.query(l - 1, r - 1))
    

File: cpp/dataStructure/sparseTable.cpp
#include <bits/stdc++.h>
using namespace std;

struct ST {
    vector<vector<int>> st;
    ST (const vector<int>& arr) {
        int n = arr.size();
        st.assign(__lg(n) + 1, vector<int>(n)); // 维度小的放前面
        for (int i = 0; i < n; ++i) st[0][i] = arr[i];
        for (int j = 1; 1 << j <= n; ++j) {
            for (int i = 0; i + (1 << j) <= n; ++i) {
                st[j][i] = max(st[j - 1][i], st[j - 1][i + (1 << (j - 1))]);
            }
        }
    }
    int query(int l, int r) { // [l, r]
        int k = __lg(r - l + 1);
        return max(st[k][l], st[k][r - (1 << k) + 1]);
    }
};

// luogu P3865 【模板】ST 表
int main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    int n, m; cin >> n >> m;
    vector<int> arr(n + 1);
    for (int i = 1; i <= n; ++i) cin >> arr[i];
    ST st(arr);
    for (int i = 0; i < m; ++i) {
        int l, r; cin >> l >> r;
        cout << st.query(l, r) << "\n";
    }
}